File: python/45_Jump_Game_II.py
# coding=utf-8
class Solution:
    # @param A, a list of integers
    # @return an integer
    # 给定一个非负整数数组，你最初位于数组的第一个位置。

    # 数组中的每个元素代表你在该位置可以跳跃的最大长度。

    # 你的目标是使用最少的跳跃次数到达数组的最后一个位置。

    # 来源：力扣（LeetCode）
    # 链接：https://leetcode-cn.com/problems/jump-game-ii
    # 著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
    def jump(self, A):
        length = len(A)
        if length<=1:
            return 0
        dist = 0# 记录当前位置能跳到的最远位置
        step = 0# 记录一共需要多少次跳跃, 一开始是0次
        lastbreak = 0# 记录上次能跳到的最远位置
        for i in range(length-1):
            # 遍历数组, 在上次能跳到的最远位置之前, 所有位置可以跳到的最远位置, 作为下一次的界限
            # 跳跃到最后位置所需要的最少步数==在最少步数内能跳到的最远范围包含最后一位
            dist = max(dist, A[i]+i)
            if i == lastbreak:
            # 当扫描已经到达了本次(step步)能到达的最远位置, 则步数+1, 边界更新为到此为止能到达的最远位置
                step += 1
                lastbreak = dist
        return step
    

File: Java/45_Jump_Game_II.java
public class Solution {
    // 写在前面：这个题的leetcode oj有问题，题目既没有说保证有解也没有说在无解的时候应该返回什么，我这里是在无解时返回-1
    // 算法：
    // 对跳跃次数贪心。每次探索在当前的跳跃次数下能到达的最远位置，如果可以到达结尾则结束，
    // 否则从当前位置到最远位置之间进行下一次跳跃，跳跃次数加1
    // 直到找到一个解或发现无解时结束
    public int jump(int[] nums) {
        if(nums==null||nums.length<=1)return 0;
        int len=nums.length;
        int cur=0,next=cur+nums[cur],candi,cnt=1;//next记录当前跳跃次数下能到达的最远位置。最初跳跃1次，能到达的最远位置是next=cur+nums[cur]
        while(next>cur&&next<len-1){
            // 进行下一次跳跃
            int run=cur+1;// 下一次跳跃途经run点
            candi = next+nums[next];// 下一次跳跃的其中一个位置
            while(run<next){// 找到直到上一次跳跃的最远点为止，能够进行的下一次跳跃所能到达的最远点
                candi = Math.max(run+nums[run], candi);
                run++;
            }
            cur=next;// 记录上一次跳跃的最远点作为下一次跳跃的搜索起点
            next=candi;// 记录下一次跳跃的最远点
            cnt++;// 跳跃次数
        }
        return next>=len-1?cnt:-1;// 判断是否有解，若无解即无法到达数列最后，则当前跳跃最远只能到达当前位置，即next==cur
    }
}